Populate metadata

will-moore opened the pull request:
As discussed https://github.com/openmicroscopy/openmicroscopy/pull/4455#issuecomment-179714477 Now that we support visualisation of bulk annotations in web (and Insight via tooltip) we want to make it easier for users to attach bulk annotations.

This PR copies the Glencoe script from https://github.com/glencoesoftware/omero-user-scripts/blob/master/util_scripts/Populate_Metadata.py and updates it a bit:
- Add (hopefully) useful description
- Use File_Annotation ID for picking files instead of File ID (web and Insight UI allow for picking File Ann)
- Make File_Annotation optional - If not set, we simply pick a .csv file from the Plate

To test, use Plate and csv files described https://github.com/openmicroscopy/ome-internal/blob/master/testing_scenarios/BulkAnnotations.txt
- Import Plate, attach test_plate.csv to it
- Select Plate and run script > Import Scripts > Populate_Metadata
- Read description. Makes sense?
- Run (don't need to choose File Annotation - will pick the csv automatically).
- Also, if multiple csv attached, you can pick the one you want from right panel and this will be used to populate the "File Annotation" parameter.
- When the script completes, selecting Wells should show the Bulk Annotation in right panel (as in https://github.com/openmicroscopy/openmicroscopy/pull/4446).

![screen shot 2016-02-04 at 15 27 54](https://cloud.githubusercontent.com/assets/900055/12819672/eab80682-cb53-11e5-8365-5787721037f5.png)

![screen shot 2016-02-04 at 15 26 12](https://cloud.githubusercontent.com/assets/900055/12819642/c67b7538-cb53-11e5-9ec1-19b10e638c60.png)

emilroz committed:
Initial Version of Populate Metadata script

emilroz committed:
Fix script behaviour when file does not exist

emilroz committed:
Extend script to screens

emilroz committed:
Handle properly screens

emilroz committed:
Fix spelling

emilroz committed:
Change plate_id to object_id

will-moore committed:
Add helpful description to Populate_Metadata.py

will-moore committed:
Use File_Annotation in params instead of File

Users can select File Annotations in the UI, but don't have easy access to File IDs

will-moore committed:
File_Annotation is optional. Choose file by name.csv

will-moore commented:
cc @emilroz I've copied your script here and made a few changes (see above).

will-moore committed:
flake8 fix

pwalczysko commented:
The output on the right-hand side is a bit repetitive. Not sure why. I have admittedly ran the script twice (deleting the ``Bulk_annotation`` attachment in between of the two script executions) but still, some lines seem to be there 4 times and more.
![screen shot 2016-02-05 at 12 01 22](https://cloud.githubusercontent.com/assets/2478303/12845826/741cd2fc-cc00-11e5-8c0e-94e84818c702.png)

pwalczysko commented:
Re: Clarified with @will-moore that the duplication came from the fact that there was another ``Bulk_annotation` attachment on the containing Screen left from previous testing. When this attachment was deleted, the duplication vanished. 

This poses an important question of ``what to do when there is multiple exemplaries of the attachment`` -> tactics might differ (even between Web and Insight), but definitely the display should be separate for each attachment in the right-hand pane display, similarly to the Key-Value pairs.

pwalczysko commented:
![screen shot 2016-02-05 at 12 31 47](https://cloud.githubusercontent.com/assets/2478303/12846583/3e788f6a-cc05-11e5-866e-31eb4ea6251c.png)

I think a hint about how do I construct the csv in case it is attached to a screen would be very useful in the description of the script (see above).
This ``screen`` workflow works for me having following .csv (see screenshot below) where the ``Plate`` column actually contains the name of the plate in OMERO (it is just misformatted by Excel as a number, see second screenshot below)


![screen shot 2016-02-05 at 12 36 58](https://cloud.githubusercontent.com/assets/2478303/12846574/306fc00a-cc05-11e5-9a58-d96e09616820.png)
![screen shot 2016-02-05 at 12 38 58](https://cloud.githubusercontent.com/assets/2478303/12846636/8c445fbc-cc05-11e5-9bb4-e786b0735a09.png)

emilroz commented:
@will-moore the script looks good to me.
@pwalczysko I think that an example `.csv` file would be better then a very long description. As you've noticed the formatting plays also very important role here, as we are really only prepared now to take properly formatted comma separated value files anything else will fail silently - you will get a single column table as an output and as a result no `bulk annotations` on the images.

pwalczysko commented:
@emilroz : Agree, an example file is better. How shouild the user find such a file ?

joshmoore commented:
Will a link to a file be clickable in web & insight? If so, I'd think a link to a [sample image](http://downloads.openmicroscopy.org/images) (perhap a fake?) which includes a CSV with a readme on how to install it would be useful?

cc: @sbesson @simleo @manics -- This would also be a first candidate for https://trello.com/c/9aEYto1f/217-table-import-prototype

will-moore commented:
@joshmoore Apologies for the delay. Unfortunately web scripts dialog doesn't make links clickable at the moment. However, I think a non-clickable link to proper docs and example csv is still the best option.
We just need to get docs in place so we have the url. Don't know if we can do this in time for this PR: @gusferguson?

gusferguson commented:
@will-moore @joshmoore - what were you thinking in terms of User Help documentation? Easy enough to have an example doc in the resources part of the User Help - that will give an immutable URL for downloads and whatever/wherever there is something in the actual User Help.

joshmoore commented:
I'll leave you guys to discuss what's doable for 5.2.2. We can capture any extras in a card for later.

NB: @will-moore, sorry for having missed it earlier, could you possibly rebase this onto of https://github.com/joshmoore/scripts/commit/47cd2ff248e3ed1649f298fa03258d33265ccd9f (i.e. replacing your 7298da80977fe7927c28accf014767938e6280e9 with commits from @emilroz so that we have the history. (I've included https://github.com/joshmoore/scripts/commit/a0dd39261f54599d9e795d6137d5cfb89c9ed599 on top to do the rename unless you want to do that yourself)

gusferguson commented:
@will-moore @joshmoore 
Confirmed that I will be putting the Help for this script in the Using Scripts page with a #metadata anchor, so the URL: 
http://help.openmicroscopy.org/scripts.html#metadata
will be the one to use.

jburel commented:
The script does not follow the approach used for ids in other scripts i.e. ids as ```rlist``` not ```rstring```
This was the source of the error in insight https://github.com/openmicroscopy/openmicroscopy/pull/4487
and it also means that the parsing will now happen in the script itself.

gusferguson commented:
Has this been tested with Insight? 
I am getting errors trying to run it on eel using today's build. 
Works fine in Web.

gusferguson commented:
Actually not running in web now.

jburel commented:
@emilroz: are you okay if we change the type for IDs i.e. list instead of String.
This will require some re-tests of the scripts
The only limitation is that the "FileAnnotation" option will potentially not work when multiple objects are selected (unless it is linked to all the objects)

jburel commented:
@gusferguson: insight PR is https://github.com/openmicroscopy/openmicroscopy/pull/4487. I pushed a fix this am (not included in the build)

gusferguson commented:
@jburel +1

will-moore committed:
Use List for IDs and improve description and message

will-moore committed:
Add #metadata to url

will-moore commented:
Updated the URL in the description as discussed.
Also switched to using List for IDs. We only support a single ID. Warning is printed to output if multiple IDs are given - we just use the first.
Added a response message to display when the script completes.

To test, select multiple Screens or Plates. Should see warning in Info when complete.
Also if completes, should see a simple message:

<img width="405" alt="screen shot 2016-02-22 at 23 26 02" src="https://cloud.githubusercontent.com/assets/900055/13236357/b222095a-d9bb-11e5-9a7d-39087ef715b6.png">

jburel commented:
Message displayed if multiple ids selected:
```
{'IDs': [226L, 8L], 'Data_Type': 'Plate'}
WARNING: Multiple IDs not currently supported
    Only using the first ID: 226
Listing files on Plate 226...
```

jburel commented:
Log in as user-4, screen ```Bulk_Annotation_test```
```WARNING:omero.util.populate_metadata:PlateColumn is unimplemented```

jburel commented:
Warning introduced in https://github.com/joshmoore/openmicroscopy/commit/00d89c64b22bd8e6adaed09f2d806861c3a5bb25
cc @joshmoore 
Web indicated that the result has been created. No message in insight only the error button.
This will have to be improved post release in insight

jburel commented:
The script works as expected now. I will retest of insight w/o https://github.com/openmicroscopy/openmicroscopy/pull/4487  included (changes no longer required)

emilroz commented:
@jburel I'm with changing the parameter type.

jburel commented:
@emilroz: Thanks, that's the best option. I have closed my insight PR.

will-moore commented:
@joshmoore That's the rebasing done that you asked for in https://github.com/ome/scripts/pull/107#issuecomment-187079776, including the renaming commit. I checked that the script is unchanged from before, (we just have the commit history) so it doesn't need retesting.

jburel commented:
Tested w/o insight changes included, ti works fine.

joshmoore commented:
Note: without the `git mv` this is now the only script we ship outside of `omero`.

will-moore commented:
Sorry @joshmoore I had cherry-picked your move commit locally but hadn't pushed it. Done now.
Ooops - I see it's just been merged already.

will-moore commented:
I'll open a new PR with the move commit.
```
* 1155250 (HEAD -> Populate_Metadata, gh/Populate_Metadata) Rename from util_scripts
```

jburel commented:
@will-moore: A PR has now been opened.

will-moore commented:
OK - e-mail delay is annoying! Closed https://github.com/ome/scripts/pull/110

joshmoore commented:
Sorry, @will-moore. I should have commented here (I had even thought about it!)
